### ISSUE TITLE
out_of_bounds_indexing doesn't apply to array references

### ISSUE THREAD

**Issue by TetroGem** [rust-lang/rust-clippy#11762]
### Summary

Array references with known length are not flagged when indexed out of bounds

### Lint Name

out_of_bounds_indexing

### Reproducer

The following correctly is flagged by clippy:
```rust
fn foo(nums: [u32; 3]) {
  nums[3]; // clippy: index out of bounds error
}
```
But the following is not:
```rust
fn foo(nums: &[u32; 3]) {
  nums[3]; // no error
}
```
Even though both have known sizes at compile time

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-pc-windows-msvc
release: 1.73.0
LLVM version: 17.0.2
```
